src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,148 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> first;
+    private Node<T> last;
+
     @Override
     public void add(T value) {
+        Node<T> lastNode = last;
+        Node<T> newNode = new Node<>(last, value, null);
+        last = newNode;
+        if (first == null) {
+            first = newNode;
+        } else {
+            lastNode.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, first);
+            first.prev = newNode;
+            first = newNode;
+            size++;
+            return;
+        }
+        Node<T> elementByIndex = findElementByIndex(index);
+        Node<T> newNode = new Node<>(elementByIndex.prev, value, elementByIndex);
+        elementByIndex.prev.next = newNode;
+        elementByIndex.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findElementByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = findElementByIndex(index).value;
+        findElementByIndex(index).value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> elementByIndex = findElementByIndex(index);
+        unlink(elementByIndex);
+        size--;
+        return elementByIndex.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = first;
+        for (int i = 0; i < size; i++) {
+            if ((node.value != null && node.value.equals(object)) || node.value == object) {
+                unlink(node);
+                size--;
```

Comment:
size--; is part of unlink logic. Is it possible to unlink node without size change?

Reply:
Looks like we don't need the method "remove" at all.